client/components/webpack-build-monitor/index.tsx

```diff
@@ -0,0 +1,153 @@
+/**
+ * @see https://webpack.js.org/api/module-variables/#__webpack_hash__-webpack-specific
+ */
+declare const __webpack_hash__: string;
+
+/**
+ * External dependencies
+ */
+import React, { FunctionComponent, useEffect, useState } from 'react';
+import classNames from 'classnames';
+import debugFactory from 'debug';
+
+/**
+ * Internal dependencies
+ */
+import Spinner from 'components/spinner';
+import { IntervalHandle, TimestampMS } from 'client/types';
+
+/**
+ * Style dependencies
+ */
+import './style.scss';
+
+const CONNECTION_TIMEOUT = 20 * 1000;
+const debug = debugFactory( 'calypso:webpack-build-monitor' );
+
+enum BuildState {
+	UNKNOWN,
+	IDLE,
+	BUILDING,
+	BUILT_WITH_WARNINGS,
+	ERROR,
+}
+
+const WebpackBuildMonitor: FunctionComponent = () => {
+	const [ lastHash, setLastHash ] = useState( __webpack_hash__ );
+	const [ isConnected, setIsConnected ] = useState( false );
+	const [ buildState, setBuildState ] = useState( BuildState.UNKNOWN );
+
+	useEffect( () => {
+		if ( typeof EventSource === 'undefined' ) {
+			if ( process.env.NODE_ENV !== 'production' ) {
+				// eslint-disable-next-line no-console
+				console.warn( 'Webpack build monitor disabled. No `EventSource`.' );
+			}
+		} else {
+			let lastActivity: TimestampMS;
+			let connectionTimer: IntervalHandle;
+			let source: EventSource;
+			const connect = () => {
+				debug( 'Webpack HMR connecting' );
+				source = new EventSource( '/__webpack_hmr' );
+
+				source.onopen = () => {
+					debug( 'Webpack HMR connected' );
+					setIsConnected( true );
+					lastActivity = Date.now();
+					connectionTimer = setInterval( () => {
+						if ( Date.now() - lastActivity > CONNECTION_TIMEOUT ) {
+							debug( 'Webpack HMR connection timeout. Reconnecting in %o…', CONNECTION_TIMEOUT );
+							setIsConnected( false );
+							clearInterval( connectionTimer );
+							source.close();
+							setTimeout( connect, CONNECTION_TIMEOUT );
```

Comment:
I don't fully understand this. If there were no incoming events in the last 20 seconds, the connection is immediately closed, then there is a 20 seconds pause, and then we reconnect again? Why the 20 seconds pause?

Reply:
The CRA client doesn't bother to reconnect, as closed connection almost always means that the dev server has been stopped: https://github.com/facebook/create-react-app/blob/6dec056de3c646fa1bce41acadc31641237863a0/packages/react-dev-utils/webpackHotDevClient.js#L71-L80